File: tests/clients/js/v8/probe.mjs
// Phase 5 D30 client probe (JS, @microsoft/signalr 8.0.17). Implements exactly the scenario in
// tests/Keryhe.Switchboard.CompatibilityTests/ClientProbeContract.md — kept byte-for-byte parallel
// with tests/clients/js/v10/probe.mjs so a JS-version-specific defect (the kind finding 1 found for
// .NET 8) can't hide behind a scenario difference between the two.
//
// Usage: node probe.mjs <apiBaseUrl> <transport> <protocol>

import { HubConnectionBuilder, HttpTransportType } from "@microsoft/signalr";
import { MessagePackHubProtocol } from "@microsoft/signalr-protocol-msgpack";

const completedSteps = [];

function parseTransport(value) {
  switch (value.toLowerCase()) {
    case "websockets":
      return HttpTransportType.WebSockets;
    case "sse":
      return HttpTransportType.ServerSentEvents;
    case "longpolling":
      return HttpTransportType.LongPolling;
    default:
      throw new Error(`Unknown transport '${value}'.`);
  }
}

function parseProtocol(value) {
  switch (value.toLowerCase()) {
    case "json":
      return false;
    case "messagepack":
      return true;
    default:
      throw new Error(`Unknown protocol '${value}'.`);
  }
}

function nextStep(lastCompleted) {
  const order = ["connect", "receive_push", "join_group", "invoke", "receive_group", "disconnect"];
  const index = order.indexOf(lastCompleted);
  return index < 0 || index === order.length - 1 ? "unknown" : order[index + 1];
}

function withTimeout(promise, seconds, label) {
  return Promise.race([
    promise,
    new Promise((_, reject) => setTimeout(() => reject(new Error(`timeout:${label}`)), seconds * 1000)),
  ]);
}

async function main() {
  const [apiBaseUrl, transportArg, protocolArg] = process.argv.slice(2);
  if (!apiBaseUrl || !transportArg || !protocolArg) {
    console.log("RESULT FAIL step=args reason=expected-3-arguments");
    process.exitCode = 1;
    return;
  }

  const transport = parseTransport(transportArg);
  const useMessagePack = parseProtocol(protocolArg);
  const suffix = Math.random().toString(16).slice(2, 10);
  const roomId = `probe-room-${suffix}`;

  const loginResponse = await fetch(`${apiBaseUrl}/api/auth/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: `probe-${suffix}` }),
  });
  if (!loginResponse.ok) {
    throw new Error(`login-failed:${loginResponse.status}`);
  }

  const { accessToken } = await loginResponse.json();

  const builder = new HubConnectionBuilder().withUrl(`${apiBaseUrl}/chatHub`, {
    transport,
    accessTokenFactory: () => accessToken,
  });

  if (useMessagePack) {
    builder.withHubProtocol(new MessagePackHubProtocol());
  }

  const connection = builder.build();

  const connected = new Promise((resolve) => connection.on("Connected", (id) => resolve(id)));
  const groupMessage = new Promise((resolve) => connection.on("ReceiveMessage", (msg) => resolve(msg)));

  await withTimeout(connection.start(), 15, "connect");
  completedSteps.push("connect");

  await withTimeout(connected, 10, "receive_push");
  completedSteps.push("receive_push");

  await withTimeout(connection.invoke("JoinRoom", roomId), 10, "join_group");
  completedSteps.push("join_group");

  await withTimeout(connection.invoke("SendMessage", roomId, "probe-payload"), 10, "invoke");
  completedSteps.push("invoke");

  const message = await withTimeout(groupMessage, 10, "receive_group");
  // MessagePack args are serialized contractless (no camelCase policy applied, unlike the JSON
  // hub protocol), so ChatHub.SendMessage's anonymous { From, Text, SentAt } arrives as
  // PascalCase over MessagePack but camelCase over JSON — accept either.
  const text = message.text ?? message.Text;
  if (text !== "probe-payload") {
    console.log(`RESULT FAIL step=receive_group reason=unexpected-payload:${text}`);
    process.exitCode = 1;
    await connection.stop().catch(() => {});
    return;
  }

  completedSteps.push("receive_group");

  await withTimeout(connection.stop(), 10, "disconnect");
  completedSteps.push("disconnect");

  console.log(`RESULT OK steps=${completedSteps.join(",")}`);
}

main()
  .catch((err) => {
    const step = completedSteps.length === 0 ? "connect" : nextStep(completedSteps[completedSteps.length - 1]);
    console.error(err);
    console.log(`RESULT FAIL step=${step} reason=${err.constructor.name}:${err.message}`);
    process.exitCode = 1;
  })
  .finally(() => {
    // An open WebSocket/long-polling handle keeps the event loop alive even after the RESULT
    // line is printed — without this, ProbeRunner's process-exit wait times out even on a
    // correctly-diagnosed failure (verified: this happened for the MessagePack cells above).
    process.exit(process.exitCode ?? 0);
  });

File: tests/clients/java/src/main/java/Probe.java
import com.microsoft.signalr.HubConnection;
import com.microsoft.signalr.HubConnectionBuilder;
import com.microsoft.signalr.TransportEnum;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.subjects.SingleSubject;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

// Phase 5 D30/D34 client probe (Java, com.microsoft.signalr 9.0.6). Implements exactly the
// scenario in tests/Keryhe.Switchboard.CompatibilityTests/ClientProbeContract.md, restricted to
// the cells the Java client SDK actually supports: TransportEnum has no
// SERVER_SENT_EVENTS member at all (only WEBSOCKETS/LONG_POLLING/ALL), and the "signalr" jar ships
// only GsonHubProtocol — no MessagePack hub protocol class exists in it. Both are real SDK
// limitations, not probe bugs: SSE and MessagePack cells are "not applicable" for this SDK, the
// same way SSE+MessagePack is "not applicable" for the service itself (03-protocol.md §1.5).
//
// Usage: java -cp ... Probe <apiBaseUrl> <transport> <protocol>
public class Probe {
    private static final List<String> completedSteps = new ArrayList<>();

    public static void main(String[] args) {
        try {
            run(args);
        } catch (Exception ex) {
            String step = completedSteps.isEmpty() ? "connect" : nextStep(completedSteps.get(completedSteps.size() - 1));
            ex.printStackTrace();
            System.out.println("RESULT FAIL step=" + step + " reason=" + ex.getClass().getSimpleName() + ":" + ex.getMessage());
            System.exit(1);
        }
    }

    private static void run(String[] args) throws Exception {
        if (args.length != 3) {
            System.out.println("RESULT FAIL step=args reason=expected-3-arguments");
            System.exit(1);
            return;
        }

        String apiBaseUrl = args[0];
        String transportArg = args[1].toLowerCase(Locale.ROOT);
        if ("sse".equals(transportArg)) {
            // No SERVER_SENT_EVENTS member exists on this SDK's TransportEnum at all — see the
            // class doc comment.
            System.out.println("RESULT FAIL step=connect reason=NotApplicable:java-sdk-has-no-sse-transport");
            System.exit(1);
            return;
        }

        TransportEnum transport = parseTransport(transportArg);
        boolean useMessagePack = parseProtocol(args[2]);
        if (useMessagePack) {
            // No MessagePack hub protocol ships in this SDK version — see the class doc comment.
            System.out.println("RESULT FAIL step=connect reason=NotApplicable:java-sdk-has-no-messagepack-hub-protocol");
            System.exit(1);
            return;
        }

        String suffix = Long.toHexString(System.nanoTime());
        String roomId = "probe-room-" + suffix;

        String accessToken = login(apiBaseUrl, "probe-" + suffix);

        HubConnection connection = HubConnectionBuilder.create(apiBaseUrl + "/chatHub")
                .withTransport(transport)
                .withAccessTokenProvider(Single.just(accessToken))
                .build();

        SingleSubject<String> connected = SingleSubject.create();
        connection.on("Connected", id -> connected.onSuccess(id), String.class);

        SingleSubject<ChatMessage> groupMessage = SingleSubject.create();
        connection.on("ReceiveMessage", msg -> groupMessage.onSuccess(msg), ChatMessage.class);

        if (!connection.start().blockingAwait(15, TimeUnit.SECONDS)) {
            throw new RuntimeException("timeout:connect");
        }

        completedSteps.add("connect");

        connected.timeout(10, TimeUnit.SECONDS).blockingGet();
        completedSteps.add("receive_push");

        if (!connection.invoke("JoinRoom", roomId).blockingAwait(10, TimeUnit.SECONDS)) {
            throw new RuntimeException("timeout:join_group");
        }

        completedSteps.add("join_group");

        if (!connection.invoke("SendMessage", roomId, "probe-payload").blockingAwait(10, TimeUnit.SECONDS)) {
            throw new RuntimeException("timeout:invoke");
        }

        completedSteps.add("invoke");

        ChatMessage message = groupMessage.timeout(10, TimeUnit.SECONDS).blockingGet();
        if (!"probe-payload".equals(message.text)) {
            System.out.println("RESULT FAIL step=receive_group reason=unexpected-payload:" + message.text);
            System.exit(1);
            return;
        }

        completedSteps.add("receive_group");

        connection.stop().blockingAwait(10, TimeUnit.SECONDS);
        completedSteps.add("disconnect");

        System.out.println("RESULT OK steps=" + String.join(",", completedSteps));
    }

    private static String login(String apiBaseUrl, String username) throws Exception {
        HttpClient http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
        String body = "{\"username\":\"" + username + "\"}";
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiBaseUrl + "/api/auth/login"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .timeout(Duration.ofSeconds(10))
                .build();

        HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() / 100 != 2) {
            throw new RuntimeException("login-failed:" + response.statusCode());
        }

        Matcher matcher = Pattern.compile("\"accessToken\"\\s*:\\s*\"([^\"]+)\"").matcher(response.body());
        if (!matcher.find()) {
            throw new RuntimeException("login-response-missing-accessToken");
        }

        return matcher.group(1);
    }

    private static TransportEnum parseTransport(String value) {
        switch (value) {
            case "websockets":
                return TransportEnum.WEBSOCKETS;
            case "longpolling":
                return TransportEnum.LONG_POLLING;
            default:
                throw new IllegalArgumentException("Unknown transport '" + value + "'.");
        }
    }

    private static boolean parseProtocol(String value) {
        switch (value.toLowerCase(Locale.ROOT)) {
            case "json":
                return false;
            case "messagepack":
                return true;
            default:
                throw new IllegalArgumentException("Unknown protocol '" + value + "'.");
        }
    }

    private static String nextStep(String lastCompleted) {
        List<String> order = List.of("connect", "receive_push", "join_group", "invoke", "receive_group", "disconnect");
        int index = order.indexOf(lastCompleted);
        return (index < 0 || index == order.size() - 1) ? "unknown" : order.get(index + 1);
    }

    // Mirrors ChatHub.SendMessage's anonymous push shape (From, Text, SentAt) — only the field the
    // scenario asserts on is required; Gson ignores unknown/extra JSON fields by default.
    public static class ChatMessage {
        public String from;
        public String text;
    }
}
